Two New FRCs: Wallet-Signing User Stories and EIP-191-equivalent-and-extension (references the user stories) (#1127)
filecoin-project/FIPs: FRCs/frc-wallet-signing-envelope.md

---
fip: "<to be assigned>"
title: Wallet-Signing Envelope for Arbitrary Messages
author: "Hugo Dias (@hugomrdias), Bumblefudge (@bumblefudge)"
discussions-to: https://github.com/filecoin-project/lotus/discussions/12761
status: Draft
type: "FRC"
---

# FRC-0XXX: Wallet-Signing Envelope for Arbitrary Messages

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the FIP.-->

> Signing arbitrary, non-transaction messages with a Filecoin or FeVM wallet opens a number of attack vectors and error cases, such as malicious message formation or non-transactions being submitted as transactions.

Reviewer: ```suggestion
Signing arbitrary, non-transaction messages with a Filecoin or FEVM wallet opens a number of attack vectors and error cases, such as malicious message formation or non-transactions being submitted as transactions.
```

A simple envelope is proposed to minimize these risks and align usage patterns with other ecosystems, which has the added benefit of making decontextualized messages easier to recontextualize.

## Abstract
<!--A short (~200 words) description of the technical issue being addressed.-->

We propose reusing the Ethereum Virtual Machine's envelope for arbitrary signing as defined in [EIP-191], and adding "version bytes" to the registry defined there for new contexts specific to the FVM.

> For context on the use-cases targeted by these variations, see [FRC-XXX].

Reviewer: is this meant to be a self-reference to this doc or some other FRC?

## Motivation
<!--The motivation is critical for FIPs that want to change the Filecoin protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the FIP solves. FIP submissions without sufficient motivation may be rejected outright.-->

> 

Reviewer: ```suggestion
Current practices can expose users to security risks such as malicious message formation and incorrect classification of non-transaction data as transactions. The proposed wallet-signing envelope aims to mitigate these issues by standardising the signing process.

By aligning with standards from other ecosystems, like the Ethereum Virtual Machine, we enhance cross-platform compatibility and reduce the learning curve for developers and users. The introduction of a versioning system for signing envelopes also ensures adaptability to future requirements, supporting a wider range of decentralised applications while maintaining robust security and user trust.
```

how about that for a motivation?

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any current Filecoin implementations. -->

(Adapted from, and extending, the earlier [ERC-191])

We propose the following format for `signed_data` interfaces:

```
0x19 <1 byte version> <version specific data> <data to sign>.
```

> The initial `0x19` byte is intended to ensure that the `signed_data` will not be mistaken for a transaction and can be "sniffed" by its first byte like a [multiformats] object or a decontextualized object from other VMs.

Reviewer: ```suggestion
The initial `0x19` byte is intended to ensure that the `signed_data` will not be mistaken for a transaction and can be "sniffed" by its first byte like a [multiformats] object or a decontextualized signature. Note that `0x19` is not valid RLP encoding, which helps prevent ambiguity and misuse.
```

Using `0x19` thus makes it possible to extend the scheme by defining a version `0x45` (`E`) to handle these kinds of signatures.

### Registry of version bytes

Note: this extends the earlier [EIP-191] registry defined in the Ethereum community

| Version byte | Specification  | Description
| ------------ | -------------- | -----------
|    `0x00`    | [EIP-191]      | Arbitary payload with intended-audience prefix

> |    `0x01`    | [eip-712]      | Structured data compatible with EVM execution
> |    `0x45`    | [eip-191]      | `personal_sign` messages from EVM connections

Reviewer: ```suggestion
|    `0x01`    | [EIP-712]      | Structured data compatible with EVM execution
|    `0x45`    | [EIP-191]      | `personal_sign` messages from EVM connections
```

|    `0x46`    | You are here   | `personal_sign` messages from FVM connections

#### Version `0x00`

```
0x19 <0x00> <intended validator address> <data to sign>
```

The version `0x00` has only `<intended validator address>` for the version specific data. In the case of a Multisig wallet that perform an execution based on a passed signature, the validator address is the address of the Multisig itself. The data to sign could be any arbitrary data.

> In ethereum contexts, there is no `chainId` affordance at the envelope layer, and little in the way of adoption.

Reviewer: ```suggestion
In Ethereum contexts, there is no `chainId` affordance at the envelope layer, and little in the way of adoption.
```

#### Version `0x45` (E)

```
0x19 <0x45 (E)> <thereum Signed Message:\n" + len(message)> <data to sign>
```

The version `0x45` (E) has `<thereum Signed Message:\n" + len(message)>` for the version-specific data. The data to sign can be any arbitrary data.

> NB: The `E` in `Ethereum Signed Message` refers to the version byte 0x45. The character `E` is `0x45` in hexadecimal which makes the remainder, `thereum Signed Message:\n + len(message)`, the version-specific data.

Using this version byte in FVM contexts is discouraged, but in FEVM contexts, it may be used safely, as FEVM messages and signatures are entirely isomorphic to the EVM.

Note that in the base `0x45` structure defined in [ERC-191], there is no envelope-level affordance for including a `chainId` (314 for Filecoin mainnet, 314159 for Filecoin testnet, 314XXX for L2s and private FEVM networks) to protect against crosschain replay attacks.

> For this reason, it is recommended to include this safety mechanism (and others like domain-binding or TTL or nonce) inside the payload, as is required by the standard [Sign-In with Ethereum][ERC-4361] template.

Reviewer: this raises the question for me that as long as you're defining a new `0x46`, why not also add this in since you're starting from scratch? `0x19 0x45 <ilecoin Signed Message:\n" + len(message) + chainId> <data to sign>`

Author: for compatibility reasons! Someone using the Metamask Snap or any other ethereum wallet that simply adds the FEVM chainId would produce wallets with the classic  `thereum Signed Message` prefix.  Unless I misunderstood the question?

Reviewer 2: i think @rvagg is saying that for native filecoin we dont need to follow strictly 191 and can just add the `chainId` in the there.

on the other hand we only really have mainnet and testnet, all the others are FEVM which need to follow strict 191 for compat.

Author: yeah it's a good point.  if no one has a counterargument, I will update the spec to match the above
